README.md
Force alias resolution in NSSavePanels under 10.7 for
recalcitrant applications. Such as InDesign. And also
10.8 (Mountain Lion)

This addresses the problems seen in http://forums.adobe.com/thread/950652.

This is beta-quality[*] software.

Install IDAlias.dylib in ~/Library

Versions are OS-dependent  
10.7	[IDAlias.dylib](https://github.com/johnhawkinson/IDAlias/blob/master/dist/0.0.1/IDAlias.dylib?raw=true)  
10.8    [IDAlias.dylib](https://github.com/johnhawkinson/IDAlias/blob/master/dist/0.0.1/IDAlias.dylib?raw=true)  
10.9    [IDAlias.dylib](https://github.com/johnhawkinson/IDAlias/blob/master/dist/0.0.2/IDAlias.dylib?raw=true)  
10.10    [IDAlias.dylib](https://github.com/johnhawkinson/IDAlias/blob/master/dist/0.0.2/IDAlias.dylib?raw=true)  

and then run

    launchctl setenv DYLD_INSERT_LIBRARIES ~/Library/IDAlias.dylib

and restart InDesign.

This method of installation will affect other apps, which might not be
wise.

For 10.11, life is trickier. Use IDAlias.dylib](https://github.com/johnhawkinson/IDAlias/blob/master/dist/0.0.2/IDAlias.dylib?raw=true) but you cannot set DYLD_INSERT_LIBRARIES globally. Instead, launch ID from the Terminal with

  DYLD_INSERT_LIBRARIES=~/Library/IDAlias.dylib '/Applications/Adobe InDesign CC 2015/Adobe InDesign CC 2015.app/Contents/MacOS/Adobe InDesign CC 2015'

More thoughts on this soon (April 2016).


---
[*] It was "alpha-quality" for 3 years and no one complained, so I guess now
it can be "beta-quality."


CHANGES
Version 0.0.2	21 April 2016
    Call [NSArray mutableCopy] since otherwise we fail on >= 10.9.
I guess [NSSavePanel URLs] used to return a mutable array and stopped.

Version 0.0.1	22 April 2013
    If we're foolish enough to set DYLD_INSERT_LIBRARIES globally, we had
better make sure to only load this code when ID or IC are running. In fact,
that's worth doing anyhow, since we don't want to run on InDesign's child
processes, like the Adobe CrashReportr.

Version 0.0.0	21 April 2013
    Initial revision.



idalias.m
#include <stdio.h>
#include <objc/Object.h>
#include <objc/runtime.h>
#include <Foundation/Foundation.h>
#include <AppKit/AppKit.h>
 
static IMP NSSavePanel_URLs = NULL;
 
@interface IDAlias:NSObject
{
}
@end

@implementation IDAlias
 
+(void)load
{
    Class originalClass = NSClassFromString(@"NSSavePanel");
    Method originalMeth = class_getInstanceMethod(originalClass, @selector(URLs));
    NSSavePanel_URLs = method_getImplementation(originalMeth);
     
    Method replacementMeth = class_getInstanceMethod(NSClassFromString(@"IDAlias"), @selector(URLs));
    method_exchangeImplementations(originalMeth, replacementMeth);
}
 
-(NSArray*)URLs
{
  int count;
  NSError *error = nil;
  NSMutableArray *urlArray;

    // We first call the original method
    urlArray = NSSavePanel_URLs(self, @selector(URLs), self);
     
    // Run our custom code
    if ((count = [urlArray count])) {
      int i;

      for (i=0; i<count; i++) {
	NSURL *url = [urlArray objectAtIndex:i];
#ifdef DEBUG
	fprintf(stderr, "URL:  %s\n",
		[[url path] UTF8String]);
#endif

	
	NSData *alias = [NSURL bookmarkDataWithContentsOfURL:url error:&error];
	if (alias == NULL) {
	  // Not an alias! A-OK!
	  continue;
	}

	BOOL isStale;
	NSURL *realURL = [NSURL URLByResolvingBookmarkData:alias options:0
					     relativeToURL:nil
				       bookmarkDataIsStale:&isStale
						     error:&error];
	if (isStale || realURL == NULL) {
	  NSLog(@"IDAlias failed alias resolution: %@", error);
	}

	[urlArray replaceObjectAtIndex:i withObject:realURL];
#ifdef DEBUG
	  NSLog(@"IDAlias replaced path with realPath:%@", [realURL path]);
#endif
      }

    }
    return urlArray;
}
 
@end


CalculatorOverrides.m
// from http://blog.timac.org/?p=761
// Simple code injection using DYLD_INSERT_LIBRARIES
// Posted: December 18th, 2012
// Author: Timac aka Alexandre Colucci
// Filed under: code injection, Debugging, DYLD_INSERT_LIBRARIES, MacOSX, Programming
// | No Comments

/* #import "ACCalculatorOverrides.h" */
 
#include <stdio.h>
#include <objc/runtime.h>
#include <Foundation/Foundation.h>
#include <AppKit/AppKit.h>
 
static IMP sOriginalImp = NULL;
 
@implementation ACCalculatorOverrides
 
+(void)load
{
    // We replace the method -[CalculatorController showAbout:] with the method -[ACCalculatorOverrides patchedShowAbout:]
    Class originalClass = NSClassFromString(@"CalculatorController");
    Method originalMeth = class_getInstanceMethod(originalClass, @selector(showAbout:));
    sOriginalImp = method_getImplementation(originalMeth);
     
    Method replacementMeth = class_getInstanceMethod(NSClassFromString(@"ACCalculatorOverrides"), @selector(patchedShowAbout:));
    method_exchangeImplementations(originalMeth, replacementMeth);
}
 
-(void)patchedShowAbout:(id)sender
{
    // We first call the original method to display the original About Box
    sOriginalImp(self, @selector(showAbout:), self);
     
    // Run our custom code which simply display an alert
    NSAlert *alert = [NSAlert alertWithMessageText:@"Code has been injected!" defaultButton:@"OK" alternateButton:nil otherButton:nil informativeTextWithFormat:@"The code has been injected using DYLD_INSERT_LIBRARIES into Calculator.app"];
    [alert runModal];
}
 
@end


IDAlias.m
// IDAlias.m
// Force alias resolution in NSSavePanels under 10.8 for
// recalcitrant applications. Such as InDesign. Only load
// ourselves in apps on our list (InDesign and InCopy).
// 
// John Hawkinson <jhawk@mit.edu>
// 21 April 2013
// 
// Derived from an example by Alexandre Colucci (timac.org)

#include <stdio.h>
#include <objc/Object.h>
#include <objc/runtime.h>
#include <Foundation/Foundation.h>
#include <AppKit/AppKit.h>
 
static IMP NSSavePanel_URLs = NULL;
 
@interface IDAlias:NSObject
{
}
@end

@implementation IDAlias
 
+(void)load
{
    NSAutoreleasePool *pool = [NSAutoreleasePool new];

    NSArray *appsToUse = [NSArray arrayWithObjects:
				  @"Adobe InDesign",
				  @"Adobe InCopy",
				  nil];
    
    /* Of the methods discussed in 
     * Technical Q&A QA1544
     * Obtaining the localized application name in Cocoa
     * http://developer.apple.com/library/mac/#qa/qa1544/_index.html
     *
     * only methods (3) and (5) return useful strings for InDesign at this
     * stage; the others return null:
     *
     * (lldb) po [[NSBundle mainBundle] bundlePath]
     * /Applications/Adobe InDesign CC 2015/Adobe InDesign CC 2015.app
     *
     * (lldb) po [[NSProcessInfo processInfo] processName]
     * Adobe InDesign CC 2015
     */

    NSString *procName = [[NSProcessInfo processInfo] processName];

    NSEnumerator *i = [appsToUse objectEnumerator];
    id id;
    BOOL found = NO;
    while ((id = [i nextObject])) {
	if ([procName rangeOfString:id].location == 0)
	    found = YES;
    }

    if (!found) {
#if DEBUG
	NSLog(@"IDAlias in proc <%@> not ID/IC. Punting.", procName);
#endif
	return;
    }

    NSLog(@"IDAlias interposing -[NSSavePanel URLs] in process <%@>.",
	  procName);

    Class originalClass = NSClassFromString(@"NSSavePanel");
    Method originalMeth = class_getInstanceMethod(originalClass,
						  @selector(URLs));
    NSSavePanel_URLs = method_getImplementation(originalMeth);
     
    Method replacementMeth = class_getInstanceMethod(
			         NSClassFromString(@"IDAlias"),
				 @selector(URLs));
    method_exchangeImplementations(originalMeth, replacementMeth);

    [pool drain];
}
 
-(NSArray*)URLs
{
  int count;
  NSError *error = nil;
  NSArray *originalURLArray;

    // We first call the original method
    originalURLArray = NSSavePanel_URLs(self, @selector(URLs), self);
    count = [originalURLArray count];

    if (!count) { // skip out early
      return originalURLArray;	
    }
    
    // Run our custom code

    NSMutableArray *urlArray = [originalURLArray mutableCopy];

    int i;

    for (i=0; i<count; i++) {
      NSURL *url = [urlArray objectAtIndex:i];
#ifdef DEBUG
      fprintf(stderr, "URL:  %s\n",
	      [[url path] UTF8String]);
#endif

	
      NSData *alias = [NSURL bookmarkDataWithContentsOfURL:url error:&error];
      if (alias == NULL) {
	// Not an alias! A-OK!
	continue;
      }
      
      BOOL isStale;

      // If we could depend on 10.10, we would use
      // [NSURL URLByResolvingAliasFileAtURL] and skip the whole
      // create bookmark/ resolve bookmark nonsense.
      NSURL *realURL = [NSURL URLByResolvingBookmarkData:alias options:0
					   relativeToURL:nil
				     bookmarkDataIsStale:&isStale
						   error:&error];
      if (isStale || realURL == NULL) {
	//	if (realURL == NULL) {
	NSLog(@"IDAlias failed alias resolution: %@", error);
      }
      
      [urlArray replaceObjectAtIndex:i withObject:realURL];
#ifdef DEBUG
      NSLog(@"IDAlias replaced path with realPath:%@", [realURL path]);
#endif
    }
    
    return urlArray;
}
 
@end
